presto-main/src/main/java/com/facebook/presto/type/DecimalToDecimalCasts.java

```diff
@@ -118,44 +119,39 @@ public static Slice shortToLongCast(
             long resultPrecision,
             long resultScale)
     {
-        return encodeUnscaledValue(bigintToBigintCast(BigInteger.valueOf(value), sourcePrecision, sourceScale, resultPrecision, resultScale));
+        return longToLongCast(unscaledDecimal(value), sourcePrecision, sourceScale, resultPrecision, resultScale);
```

Comment:
The problem is that if we want to have reusable arithmetic we have to have some value storage (Slice in this case). That is why I opt for `result` parameter in arithmetic. We can also use accumulator `result` field as in ArrayConcatFunction.